src/algebra/group/inj_surj.lean

```diff
@@ -329,6 +355,17 @@ protected def monoid [monoid M₁] (f : M₁ → M₂) (hf : surjective f)
   npow_succ' := λ n, hf.forall.2 $ λ x, by erw [←npow, pow_succ, ←npow, ←mul],
    .. hf.semigroup f mul, .. hf.mul_one_class f one mul }
 
+/-- A type endowed with `0`, `1` and `+` is an additive monoid with one,
+if it admits a surjective map that preserves `0`, `1` and `*` from an additive monoid with one.
+See note [reducible non-instances]. -/
+@[reducible]
+protected def add_monoid_with_one {M₂} [has_zero M₂] [has_one M₂] [has_add M₂] [has_scalar ℕ M₂]
+  [add_monoid_with_one M₁] (f : M₁ → M₂) (hf : surjective f)
+  (zero : f 0 = 0) (add : ∀ x y, f (x + y) = f x + f y)
+  (nsmul : ∀ x (n : ℕ), f (n • x) = n • f x) :
+  add_monoid_with_one M₂ :=
+{ one := 1, .. hf.add_monoid f zero add nsmul }
```

Comment:
Should we use `f (n : M₁)` as `nat_cast` here?

Reply:
As I've written on zulip:

> Since #12126 has been merged now it might be good to revisit `function.injective.ring` & co.  These currently take `add_group_with_one` arguments instead of `has_zero`/`has_one`/`has_add`/`has_inv`/..., which is really inconsistent.  There is also a type class diamond issue now in `ring_theory/integral_closure` that could be related.  We should probably add `has_nat_cast`/`has_int_cast` type classes and handle the casts like any other operation in `function.injective.ring`.

So this function should really take a `[has_nat_cast M₂]` argument (and also `f 1 = 1` and `∀ n, f (nat.cast n) = nat.cast n`).

Reply:
This is implemented now.